File: day06.py
"""Solution to day 6 of Advent of Code"""

from collections import defaultdict

from get_input import get_input, line_parser


def part1(orbits):
    orbit_map = {b: a for a, b in orbits}
    orbit_count = 0
    for orbiter in orbit_map.keys():
        while orbiter in orbit_map:
            orbiter = orbit_map[orbiter]
            orbit_count += 1
    return orbit_count


def part2(orbits):
    orbit_map = defaultdict(set)
    for a, b in orbits:
        orbit_map[a].add(b)
        orbit_map[b].add(a)
    goal = list(orbit_map['SAN'])[0]
    start = list(orbit_map['YOU'])[0]
    queue = [(0, start)]
    seen = set()
    while queue:
        steps, current = queue.pop(0)
        if current == goal:
            return steps
        if current in seen:
            continue
        seen.add(current)
        for step in orbit_map[current]:
            queue.append((steps+1, step))
    raise Exception("Cannot be done")


def parse(line):
    a, b = line.split(')')
    return (a, b)


TEST1 = """COM)B
B)C
C)D
D)E
E)F
B)G
G)H
D)I
E)J
J)K
K)L"""


if __name__ == '__main__':
    assert part1(line_parser(TEST1.strip(), parse=parse)) == 42
    lines = line_parser(get_input(6, 2019), parse=parse)
    print(f"Part 1: {part1(lines)}")
    test2 = line_parser(TEST1.strip(), parse=parse)
    test2.append(("K", "YOU"))
    test2.append(("I", "SAN"))
    assert part2(test2) == 4
    print(f"Part 2: {part2(lines)}")

File: day06.rs
use common::get_input;

use std::collections::{HashMap, HashSet, VecDeque};

type Input = HashMap<String, String>;

fn part1(orbit_map: &Input) -> usize {
    let mut orbit_count = 0;
    for orbit in orbit_map.keys()  {
        let mut orbiter = orbit;
        while orbit_map.contains_key(orbiter) {
            orbiter = orbit_map.get(orbiter).unwrap();
            orbit_count += 1;
        }
    }
    orbit_count
}

fn part2(orbits: &Input) -> usize {
    let mut orbit_map = HashMap::new();
    for (a, b) in orbits {
        orbit_map.entry(a).or_insert(HashSet::new()).insert(b);
        orbit_map.entry(b).or_insert(HashSet::new()).insert(a);
    }
    let goal = orbit_map.get(&"SAN".to_string()).unwrap().iter().next().unwrap();
    let start = orbit_map.get(&"YOU".to_string()).unwrap().iter().next().unwrap();
    let mut queue = VecDeque::new();
    let mut seen: HashSet<String> = HashSet::new();
    queue.push_back((0, start));
    while !queue.is_empty() {
        let (steps, current) = queue.pop_front().unwrap();
        if current == goal {
            return steps
        }
        if seen.contains(*current) {
            continue
        }
        seen.insert((*current).clone());
        for step in orbit_map.get(current).unwrap() {
            queue.push_back((steps+1, step));
        }
    }
    panic!("Unable to find a solution");
}

fn parse(text: &str) -> Input {
    let mut orbit_map = HashMap::new();
    for line in text.trim().split('\n') {
        let mut parts = line.trim().split(')');
        let a = parts.next().unwrap();
        let b = parts.next().unwrap();
        assert!(parts.next().is_none());
        orbit_map.insert(b.to_string(), a.to_string());
    }
    orbit_map
}

fn main() {
    let input = parse(&get_input(06, 2019));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
